dts: bindings: add XIP node for nordic devices

hakonfam opened the pull request:
There is currently no information available about the XIP
addresses for the nRF52840 and nRF5340.

Add a new node type and add the nodes to the qspi bus.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

mbolivar-nordic commented:
@galak I am curious how you think this sort of thing should be handled as I expect other SoCs will need similar or may already have similar nodes.

We are trying to model this XIP area of a memory map:

![Screenshot from 2021-08-31 13-24-21](https://user-images.githubusercontent.com/62298616/131571045-9deeefec-5193-4835-9eb6-80a0b7ec0b21.png)

This is memory mapped external flash memory and the flash controller in this case is the nordic,nrf-qspi node.

galak commented:
@mbolivar-nordic at a quick glance this seems similar to some other QSPI devices that expose a memory map XIP region.  We use the reg property of the qspi and have one of the reg's map to the XIP region.

Here's an example:
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/spi/ti_qspi.txt

galak commented on dts/bindings/mtd/nordic,nrf-qspi-xip.yaml:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  FOR INTERNAL USE ONLY
```
what is this meant to me?

galak reviewed (changes requested):
Need to fix description in binding.

galak commented:
Is it really the case that there is 1 byte that isn't part of the XIP region?  That seems very weird.

mbolivar-nordic commented:
> Need to fix description in binding.

Thanks, yeah. We probably should have tagged this RFC for now; it was more a question about how to represent this than about the details in the description.

mbolivar-nordic commented:
> the reg property of the qspi and have one of the reg's map to the XIP region.
> 
> Here's an example:
> https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/spi/ti_qspi.txt

That looks clean, especially with the reg names.

mbolivar-nordic committed:
cmake: extensions: add NAME parameter to DT register helpers

Allow register address and size access by name.

Example devicetree fragment:

  / {
          foo@deadbeef {
                  reg = <0xdeadbeef 0x1000>, <0xfeedface 0x2000>;
                  reg-names = "bar", "baz";
          };
  };

Example usage:

  dt_reg_addr(bar_addr PATH "/foo@deadbeef" NAME bar)
  dt_reg_size(bar_size PATH "/foo@deadbeef" NAME bar)

  dt_reg_addr(baz_addr PATH "/foo@deadbeef" NAME baz)
  dt_reg_size(baz_size PATH "/foo@deadbeef" NAME baz)

Results:

  - bar_addr is 0xdeadbeef
  - bar_size is 0x1000
  - baz_addr is 0xfeedface
  - baz_size is 0x2000

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@galak I think this is what you meant; right? I had to extend the cmake helpers to make this useful.

mbolivar-nordic commented:
@tejlmand @hakonfam example:

```cmake
dt_nodelabel(qspi_path NODELABEL "qspi")

dt_reg_addr(qspi_addr PATH ${qspi_path} NAME qspi)
dt_reg_size(qspi_size PATH ${qspi_path} NAME qspi)
dt_reg_addr(qspi_mm_addr PATH ${qspi_path} NAME qspi_mm)
dt_reg_size(qspi_mm_size PATH ${qspi_path} NAME qspi_mm)

message(STATUS "qspi_addr=${qspi_addr}")
message(STATUS "qspi_mm_addr=${qspi_mm_addr}")
message(STATUS "qspi_size=${qspi_size}")
message(STATUS "qspi_mm_size=${qspi_mm_size}")
```

output on nRF52840:

```
-- qspi_addr=0x40029000
-- qspi_mm_addr=0x12000000
-- qspi_size=0x1000
-- qspi_mm_size=0xdffffff
```

> Is it really the case that there is 1 byte that isn't part of the XIP region? That seems very weird.

@hakonfam could you check on this?

mbolivar-nordic commented:
> cmake: extensions: fix copy/paste issues in DT error messages

I split this part into a separate PR for LTS: https://github.com/zephyrproject-rtos/zephyr/pull/38185

hakonfam commented:
> Is it really the case that there is 1 byte that isn't part of the XIP region? That seems very weird.

I fixed that.

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2760,6 +2780,17 @@ function(dt_reg_size var)
 endfunction()
 
+# Internal helper for dt_reg_addr/dt_reg_size; not meant to be used directly
+function(_dt_reg_get_index var path name)
```
Don't use `_` as first character of a function.
The `_` is used by CMake to prepend existing functions when they are overloaded, as can be seen by this example:
```
# test.cmake
function(A)
  message("I'm in A")
endfunction()

function(A)
  message("I'm in B")
endfunction()

A()
_A()
```
and running this:
```
$ cmake -P test.cmake
I'm in B
I'm in A
```

mbolivar-nordic replied:
> `_` is used by CMake to prepend existing functions when they are overloaded

Cool. I didn't know that. Fixed now.

tejlmand reviewed (changes requested):
Please, don't use `_` as first character in a function name, that's bad habit in CMake.

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2760,6 +2780,17 @@ function(dt_reg_size var)
+  else()
+    list(FIND reg_names "${name}" index)
+  endif()
+  set("${var}" "${index}" PARENT_SCOPE)
```
Why the quotes on the return var ?

If there is any risk that `${var}` can be undefined or empty string, then we should not try to do a `set()` on it, but based on how this is used, that is not really a risk.
```suggestion
  set(${var} "${index}" PARENT_SCOPE)
```

tejlmand replied:
@mbolivar-nordic why was this marked resolved without any comments or fixes ?

mbolivar-nordic commented:
Done

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2788,7 +2788,7 @@ function(dt_reg_index_private var path name)
     list(FIND reg_names "${name}" index)
   endif()
-  set("${var}" "${index}" PARENT_SCOPE)
+  set(${var} "${index}" PARENT_SCOPE)
```
ahh, here's the reason.
I believe this should be in previous commit.

mbolivar-nordic replied:
> I believe this should be in previous commit.

@tejlmand You're right, of course. Fixed.

mbolivar-nordic committed:
dts: nordic: add QSPI XIP area registers for nordic devices

There is currently no information available about the XIP
addresses for the nRF52840 and nRF5340.

Add this via a new 'reg' block in the relevant QSPI nodes, along with
names for the register blocks so they are easier to retrieve.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

tejlmand reviewed (approved):
lgtm.

mbolivar-nordic commented:
Rebased

cfriedt commented:
@galak - please go ahead and merge if the requested changes have been made (quite likely, I think).
